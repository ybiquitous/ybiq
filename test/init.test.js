const path = require("path");
const os = require("os");
const fs = require("fs");
const fse = require("fs-extra");
const test = require("tape");
const pkg = require("../package.json");
const init = require("../lib/init");
const exec = require("./helpers/exec");

/* eslint-disable node/no-unsupported-features/node-builtins */
const readFile = (file) => fs.promises.readFile(file, "utf8");
const readJSON = (file) => fs.promises.readFile(file, "utf8").then(JSON.parse);
/* eslint-enable node/no-unsupported-features/node-builtins */

const sandbox = async (fn, t) => {
  const workDir = path.join(os.tmpdir(), `${pkg.name}${process.hrtime.bigint()}`);
  await fs.promises.mkdir(workDir); // eslint-disable-line node/no-unsupported-features/node-builtins

  const logMsgs = [];
  const logger = (msg) => logMsgs.push(msg);

  try {
    const cwd = process.cwd();
    const fixturePath = (name) => path.join(cwd, "test", "fixtures", name);
    const fixture = async (name) => {
      const src = fixturePath(name);
      const dest = path.join(workDir, "package.json");
      await fs.promises.copyFile(src, dest); // eslint-disable-line node/no-unsupported-features/node-builtins
      return dest;
    };

    return await fn(t, {
      fixturePath,
      fixture,
      readFixture: (name) => readFile(fixturePath(name)),
      readFixtureJSON: (name) => readJSON(fixturePath(name)),
      readOrigFile: (name) => readFile(path.join(cwd, name)),
      readWorkFile: (name) => readFile(path.join(workDir, name)),
      logMessage: () => logMsgs.join(""),
      initArgs: { cwd: workDir, logger },
    });
  } finally {
    // TODO: Node 12+
    // await fs.promises.rmdir(workDir, { recursive: true });
    await fse.remove(workDir);
  }
};

test("init", (t) => {
  const testInSandbox = (name, fn) => {
    t.test(name, (t) => sandbox(fn, t));
  };

  testInSandbox('update "package.json"', async (t, ctx) => {
    const src = await ctx.fixture("package-normal.json");
    await init(ctx.initArgs);
    const actual = await readJSON(src);
    const expected = await ctx.readFixtureJSON("package-normal_expected.json");
    t.deepEqual(actual, expected);
    t.end();
  });

  testInSandbox('update "package.json" without fields', async (t, ctx) => {
    const src = await ctx.fixture("package-empty.json");
    await init(ctx.initArgs);
    const actual = await readJSON(src);
    const expected = await ctx.readFixtureJSON("package-empty_expected.json");
    t.deepEqual(actual, expected);
    t.end();
  });

  [
    ".editorconfig",
    ".remarkignore",
    ".github/workflows/commitlint.yml",
    ".github/workflows/npm-audit-fix.yml",
    ".github/workflows/release.yml",
    ".github/workflows/test.yml",
  ].forEach((file) => {
    testInSandbox(`write "${file}"`, async (t, ctx) => {
      await ctx.fixture("package-normal.json");
      await init(ctx.initArgs);
      t.ok(ctx.logMessage().includes("`package.json` was updated"));

      const original = await ctx.readOrigFile(file);
      const copy = await ctx.readWorkFile(file);
      t.is(original, copy);

      t.ok(pkg.files.includes(file));
      t.end();
    });
  });

  testInSandbox("throw error if no package.json", async (t, ctx) => {
    const error = await init(ctx.initArgs).catch((err) => err);
    t.ok(error instanceof Error);
    t.is(error.code, "ENOENT");
    t.end();
  });

  testInSandbox("End-to-End via CLI", async (t, ctx) => {
    await ctx.fixture("package-normal.json");
    const { stdout, stderr } = await exec("init", { cwd: ctx.initArgs.cwd });
    t.match(
      stdout,
      new RegExp( // eslint-disable-line prefer-regex-literals
        `=> \`package.json\` was updated
=> \`.editorconfig\` was updated
=> \`.remarkignore\` was updated
=> \`.github/workflows/commitlint.yml\` was updated
=> \`.github/workflows/npm-audit-fix.yml\` was updated
=> \`.github/workflows/release.yml\` was updated
=> \`.github/workflows/test.yml\` was updated`,
        "u"
      )
    );
    t.is(stderr, "");
    t.end();
  });
});
