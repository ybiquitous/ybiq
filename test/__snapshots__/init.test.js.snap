// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update "package.json" 1`] = `
{
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional",
    ],
    "rules": {
      "body-max-line-length": [
        1,
        "always",
        100,
      ],
    },
  },
  "eslintConfig": {
    "extends": [
      "ybiquitous",
    ],
    "ignorePatterns": [
      "coverage",
      "dist",
      "tmp",
    ],
    "reportUnusedDisableDirectives": true,
    "root": true,
  },
  "lint-staged": {
    "!(*.snap)": "prettier --cache --write",
    "!(CHANGELOG).md": "remark --frail",
    "*.css": "xyz",
    "*.{js,jsx,cjs,mjs,ts,tsx}": "eslint --cache --fix",
  },
  "publishConfig": {
    "access": true,
    "provenance": true,
  },
  "remarkConfig": {
    "plugins": [
      "remark-preset-ybiquitous",
    ],
  },
  "scripts": {
    "clean": "git clean -dx --force --exclude=node_modules --exclude=.husky",
    "format": "npm-run-all --print-label --silent --parallel lint:*:fix",
    "lint": "npm-run-all --print-label --silent --parallel lint:*",
    "lint:commit": "commitlint --from HEAD~10",
    "lint:js": "eslint --cache --ext=js,jsx,cjs,mjs,ts,tsx .",
    "lint:js:fix": "npm run lint:js -- --fix",
    "lint:md": "remark . --frail",
    "lint:md:fix": "remark . --output",
    "lint:styles": "npm run prettier -- --check",
    "lint:styles:fix": "npm run prettier -- --write",
    "lint:types": "tsc --noEmit",
    "lint:types:watch": "npm run lint:types -- --watch",
    "prepare": "husky install",
    "prerelease": "git switch main && git pull && npm ci && npm run clean && npm test && npm run clean",
    "pretest": "npm run lint",
    "prettier": "prettier --cache .",
    "release": "standard-version",
    "release:dry-run": "standard-version --dry-run",
    "test": "abc",
    "test:coverage": "npm --ignore-scripts test -- --coverage",
    "test:watch": "npm --ignore-scripts test -- --watch",
  },
  "standard-version": {
    "scripts": {
      "postchangelog": "prettier --write CHANGELOG.md",
    },
    "sign": true,
  },
}
`;

exports[`update "package.json" without fields 1`] = `
{
  "commitlint": {
    "extends": [
      "@commitlint/config-conventional",
    ],
    "rules": {
      "body-max-line-length": [
        1,
        "always",
        100,
      ],
    },
  },
  "eslintConfig": {
    "extends": [
      "ybiquitous",
    ],
    "ignorePatterns": [
      "coverage",
      "dist",
      "tmp",
    ],
    "reportUnusedDisableDirectives": true,
    "root": true,
  },
  "lint-staged": {
    "!(*.snap)": "prettier --cache --write",
    "!(CHANGELOG).md": "remark --frail",
    "*.{js,jsx,cjs,mjs,ts,tsx}": "eslint --cache --fix",
  },
  "publishConfig": {
    "provenance": true,
  },
  "remarkConfig": {
    "plugins": [
      "remark-preset-ybiquitous",
    ],
  },
  "scripts": {
    "clean": "git clean -dx --force --exclude=node_modules --exclude=.husky",
    "format": "npm-run-all --print-label --silent --parallel lint:*:fix",
    "lint": "npm-run-all --print-label --silent --parallel lint:*",
    "lint:commit": "commitlint --from HEAD~10",
    "lint:js": "eslint --cache --ext=js,jsx,cjs,mjs,ts,tsx .",
    "lint:js:fix": "npm run lint:js -- --fix",
    "lint:md": "remark . --frail",
    "lint:md:fix": "remark . --output",
    "lint:styles": "npm run prettier -- --check",
    "lint:styles:fix": "npm run prettier -- --write",
    "lint:types": "tsc --noEmit",
    "lint:types:watch": "npm run lint:types -- --watch",
    "prepare": "husky install",
    "prerelease": "git switch main && git pull && npm ci && npm run clean && npm test && npm run clean",
    "pretest": "npm run lint",
    "prettier": "prettier --cache .",
    "release": "standard-version",
    "release:dry-run": "standard-version --dry-run",
    "test": "test",
    "test:coverage": "npm --ignore-scripts test -- --coverage",
    "test:watch": "npm --ignore-scripts test -- --watch",
  },
  "standard-version": {
    "scripts": {
      "postchangelog": "prettier --write CHANGELOG.md",
    },
    "sign": true,
  },
}
`;

exports[`write ".editorconfig" 1`] = `
"# editorconfig.org
root = true

[*]
end_of_line = lf
insert_final_newline = true
trim_trailing_whitespace = true
charset = utf-8
indent_style = space
indent_size = 2
max_line_length = 100

# trailing spaces in markdown indicate word wrap
[*.md]
trim_trailing_whitespace = false
"
`;

exports[`write ".github/workflows/dependabot-auto-merge.yml" 1`] = `
"name: Dependabot auto-merge

on:
  pull_request:
    types: [opened, synchronize]

jobs:
  auto-merge:
    uses: ybiquitous/.github/.github/workflows/dependabot-auto-merge-reusable.yml@main
"
`;

exports[`write ".github/workflows/npm-audit-fix.yml" 1`] = `
"name: npm audit fix

on:
  schedule:
    - cron: "0 0 * * *"
  workflow_dispatch:

jobs:
  npm-audit-fix:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      pull-requests: write
    steps:
      - uses: actions/checkout@v4
      - uses: ybiquitous/npm-audit-fix-action@v5
        with:
          github_token: \${{ secrets.PERSONAL_ACCESS_TOKEN }}
"
`;

exports[`write ".github/workflows/npm-diff.yml" 1`] = `
"name: npm diff

on:
  pull_request_target:
    types: [opened]

jobs:
  npm-diff:
    if: \${{ startsWith(github.head_ref, 'dependabot/npm_and_yarn/') }}
    runs-on: ubuntu-latest
    permissions:
      contents: read
      pull-requests: write
    steps:
      - uses: ybiquitous/npm-diff-action@v1
"
`;

exports[`write ".github/workflows/release.yml" 1`] = `
"name: Release

on:
  push:
    tags: ["**"]

jobs:
  release:
    uses: ybiquitous/.github/.github/workflows/nodejs-release-reusable.yml@main
    secrets:
      npm-token: \${{ secrets.NPM_TOKEN }}
"
`;

exports[`write ".github/workflows/test.yml" 1`] = `
"name: Test

on:
  push:
    branches: ["main"]
  pull_request:
    branches: ["**"]

jobs:
  test:
    strategy:
      fail-fast: false
      matrix:
        node-version: ["16", "18", "20"]
    uses: ybiquitous/.github/.github/workflows/nodejs-test-reusable.yml@main
    with:
      node-version: \${{ matrix.node-version }}

  lint:
    uses: ybiquitous/.github/.github/workflows/nodejs-lint-reusable.yml@main
"
`;

exports[`write ".husky/commit-msg" 1`] = `
"#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

npx --no-install commitlint --edit "$1"
"
`;

exports[`write ".husky/post-commit" 1`] = `
"#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

npm run release:dry-run
"
`;

exports[`write ".husky/pre-commit" 1`] = `
"#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

npx --no-install lint-staged
"
`;

exports[`write ".remarkignore" 1`] = `
"CHANGELOG.md
"
`;
